eva_cttv_pipeline/evidence_string_generation/annotated_clinvar.py

```diff
@@ -0,0 +1,85 @@
+import xml.etree.ElementTree as ET
+
+from eva_cttv_pipeline.clinvar_xml_io.clinvar_xml_io import ClinVarTrait, ClinVarRecordMeasure, ClinVarDataset, \
+    ClinVarRecord
+from eva_cttv_pipeline.clinvar_xml_io.clinvar_xml_io.xml_parsing import iterate_rcv_from_xml
+from eva_cttv_pipeline.evidence_string_generation.clinvar_to_evidence_strings import load_efo_mapping, \
+    get_consequence_types
+from eva_cttv_pipeline.evidence_string_generation import consequence_type as CT
+
+
+PROVIDER = 'ClinVarXmlPipeline'
+
+
+class AnnotatingClinVarDataset(ClinVarDataset):
+    """This class provides the ability to annotate ClinVar records with EFO mappings and consequence mappings."""
+
+    def __init__(self, clinvar_xml, string_to_efo_mappings, variant_to_gene_mappings):
+        super().__init__(clinvar_xml)
+        self.header = f'''<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<ReleaseSet Dated="{self.today()}" ProcessedBy="{PROVIDER}" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" Type="full" xsi:noNamespaceSchemaLocation="http://ftp.ncbi.nlm.nih.gov/pub/clinvar/xsd_public/clinvar_public_1.60.xsd">
+'''
+        self.header = self.header.encode('utf-8')
+        self.string_to_efo_mappings = string_to_efo_mappings
+        self.variant_to_gene_mappings = variant_to_gene_mappings
+
+    def __iter__(self):
+        for rcv in iterate_rcv_from_xml(self.clinvar_xml):
+            record = AnnotatedClinVarRecord(rcv)
+            self.annotate(record)
+            yield record
+
+    def annotate(self, record):
+        # Functional consequences for measure
+        if record.measure:
+            consequence_types = get_consequence_types(record.measure, self.variant_to_gene_mappings)
+            record.measure.add_ensembl_annotations(consequence_types)
+        # EFO terms for trait
+        for trait in record.traits_with_valid_names:
+            efo_ids = []
+            for trait_name in trait.all_names:
+                efo_ids.extend(efo_id for efo_id, efo_label in self.string_to_efo_mappings.get(trait_name.lower(), []))
+            trait.add_efo_mappings(efo_ids)
+
+
+class AnnotatedClinVarRecord(ClinVarRecord):
+
+    def __init__(self, rcv):
+        super().__init__(rcv, trait_class=EfoMappedClinVarTrait, measure_class=EnsemblAnnotatedClinVarMeasure)
+
+
+class EfoMappedClinVarTrait(ClinVarTrait):
+
+    def add_efo_mappings(self, efo_ids):
+        efo_elts = []
+        for efo_id in efo_ids:
+            if efo_id.startswith('http'):
+                efo_id = efo_id.split('/')[-1].replace('_', ':')
```

Comment:
Good point, in our case `string_to_efo_mappings` stores URIs (as [here](https://github.com/EBIvariation/eva-opentargets/blob/fc0303d4278f0b3a05faf57e69c2526a58f6d9bf/tests/eva_cttv_pipeline/evidence_string_generation/resources/string_to_ontology_mappings.tsv)), so if you use our trait mapping pipelines it should be internally consistent (though not consistent with what ClinVar already has, which is already not consistent, as you've noted).

On the other hand if someone produced mappings in another way and those weren't formatted consistently, this output wouldn't be either, it would I guess just translated URIs to CURIEs and leave everything else alone.  Let me know what you think, we could maybe put a more generic formatter in there.

Reply:
To understand better the issue, can you remind me if ``efo_ids`` contains exclusively the ones we annotate, or all of the ones originally in the XML? If that is the case, this syntax would be okay as it is, since, as you mention, we have consistently URIs.

I guess ``efo_ids`` are indeed only the annotated ones, or otherwise we would be losing information from ClinVar in the annotation process, and also wouldn't be able to accept other sources of annotation within the XML.

Reply:
Yes that's right, here `efo_ids` are the annotations we're providing, the other xrefs originally in ClinVar are preserved unchanged.